demo/src/__tests__/MemoryModels.spec.tsx

```diff
@@ -75,4 +66,102 @@ describe("MemoryModelsTab", () => {
         });
     });
+
+    describe("MemoryModelsFileInput", () => {
+        beforeEach(() => {
+            const textDataMock = "";
+
+            render(
+                <MemoryModelsUserInput
+                    onTextDataSubmit={onSubmitMock}
+                    setTextData={setTextDataMock}
+                    textData={textDataMock}
+                />
+            );
+        });
+
+        afterEach(() => {
+            // spies / mocks need to be manually restored to not fail subsequent tests
+            jest.restoreAllMocks();
+        });
+
+        it("renders an enabled input and disabled reapply button", () => {
+            const input: HTMLInputElement = screen.getByTestId("file-input");
+            expect(input).toHaveProperty("disabled", false);
+
+            const reapplyBtn = screen.getByTestId("file-input-reapply-button");
+            expect(reapplyBtn).toHaveProperty("disabled", true);
+        });
+
+        it("calls console error and setTextData when file upload fails", async () => {
+            const mockErrorMessage = "Mock error message";
+            jest.spyOn(global, "FileReader").mockImplementationOnce(() => {
+                throw new Error(mockErrorMessage);
+            });
+            const consoleErrorSpy = jest.spyOn(console, "error");
+
+            const file = new File(
+                [JSON.stringify({ id: 1, uuid: 2 })],
+                "test.json",
+                {
+                    type: "application/json",
+                }
+            );
+            const input: HTMLInputElement = screen.getByTestId("file-input");
+            await waitFor(() => {
+                // this needs to be awaited because of fileReader.onload being async
+                fireEvent.change(input, { target: { files: [file] } });
+            });
+
+            expect(consoleErrorSpy).toHaveBeenNthCalledWith(
+                1,
+                `Error parsing uploaded File as JSON: ${mockErrorMessage}`
+            );
+            expect(setTextDataMock).toHaveBeenNthCalledWith(1, null);
+        });
+
+        describe("when a file is uploaded", () => {
+            const fileString = JSON.stringify({ id: 1, uuid: 2 });
+            let input: HTMLInputElement;
+
+            beforeEach(async () => {
+                const file = new File([fileString], "test.json", {
+                    type: "application/json",
+                });
+                input = screen.getByTestId("file-input");
+                fireEvent.change(input, { target: { files: [file] } });
+            });
+
+            it("enables reapply Button", async () => {
+                const reapplyBtn = screen.getByTestId(
+                    "file-input-reapply-button"
+                );
+                await waitFor(() => {
+                    expect(reapplyBtn).toHaveProperty("disabled", false);
+                });
+            });
+
+            it("clicking reapply button calls setTextData", async () => {
```

Comment:
See comments for expansions, tried several approaches / combinations and this weird-ish syntax is the only one that worked.

Reply:
That looks good to me 👍